### PR TITLE
Add support for Ambit3 Run

### PR DESCRIPTION
**Pull request by soutade**
Add support for Ambit3 Run
Found 0x0b1e command (get_compact_serial) and add its dissector
Commit 3e94e19 NEEDS TESTS

### CODE REVIEW

**Comment by soutade** [comment]
It's for latest 2.4.1 firmware

File: src/libambit/device_driver_ambit3.c [hunk 1]
```diff
@@ -579,7 +579,7 @@ static int parse_log_header(libambit_sbem0102_data_t *reply_data_object, ambit3_
         data += 19;
         break;
       case 0x8a: /* gen2+gen3 fw */
-        data += 4;
+        data += 1; // Skip size ??
```
Comment (landas) [inline src/libambit/device_driver_ambit3.c]: I know this part of code is is kind of "broken", but I think this change might break the log parsing?
Line 575 - 582 is a hack making the code backward compatible with older versions of firmware.

Have you tested it to work to get activity logs from your ambit3 with newest firmware?

**Review by landas — COMMENTED** [review]
Thank you very much for this pull request. 
I have some questions, because I think it might break the log parsing.

File: src/libambit/pmem20.c [hunk 2]
```diff
@@ -93,7 +93,7 @@ int libambit_pmem20_log_init(libambit_pmem20_t *object, uint32_t mem_start, uint
 
     // Set memory structure
     object->log.mem_start = mem_start;
-    object->log.mem_size = mem_size;
+    object->log.mem_size = (mem_size) ?:-1; // mem_size is 0 ??
```
Comment (landas) [inline src/libambit/pmem20.c:96]: Could you tell me why this change is necessary?
Does it relate to commit 3e94e19 ?

**Review by landas — CHANGES_REQUESTED** [review]

File: src/libambit/device_driver_ambit3.c (hunk 1 above)
Comment (soutade) [reply src/libambit/device_driver_ambit3.c]: You're right. We have firmware generation version as a parameter. I'll send you another pull request to fix it and be backward compatible.

File: src/libambit/pmem20.c (hunk 2 above)
Comment (soutade) [reply src/libambit/pmem20.c:96]: The response is in the message description. We initialize log.mem_size with Exercise log size (got from memory map), and checks for size overflow, but it's 0 on my version which lead parser to fail.

File: src/libambit/device_driver_ambit3.c [hunk 3]
```diff
@@ -579,7 +579,10 @@ static int parse_log_header(libambit_sbem0102_data_t *reply_data_object, ambit3_
         data += 19;
         break;
       case 0x8a: /* gen2+gen3 fw */
-        data += 1; // Skip size ??
+        if (fw_gen == AMBIT3_FW_GEN3)
+            data += 1; // Skip size
```
Comment (landas) [inline src/libambit/device_driver_ambit3.c]: Hmm.. I understand where you want to go with this, and I agree it is the correct way. But my guess is it will break the parse_log_header for AMBIT_FW_GEN3, because the src/libambit/device_driver_ambit3.c is not perfect.

The 0x8a type starts with:
8aff 7e03 0000 [Here is the date/time]

libambit_sbem0102_data_ptr(data) -> returns data[2]:
7e03 0000 [Here is the date/time]

So we skip first 4 bytes to get to date/time.

And to know where to end the log_head we use some super magic (ironic) code.
It is not perfect, but it works without rewriting sbem0102.

I want you to reverse commit  	3e94e19 to d14b874, not because they are wrong, but it will need a total rewrite of the parser code to not break the current logic. (as of my understanding of the device_driver_ambit3.c code)

**Review by landas — CHANGES_REQUESTED** [review]

File: src/libambit/device_driver_ambit3.c (hunk 3 above)
Comment (soutade) [reply src/libambit/device_driver_ambit3.c]: Done in pull request #2. I keep these modifications on my branch if some people just wants it to work for now.

In my dump, I only have 0x8a [size on one byte] [time] (eg : 0x8a 0x95 2017-04-29T1...) . You can see it in [monambit-openambit.zip](https://github.com/landas/openambit/files/967960/monambit-openambit.zip)

It only contains a dummy activity because I upgraded my firmware before finalizing code changes.

Comment (landas) [reply src/libambit/device_driver_ambit3.c]: Thanks for your usb dump. I have to look more into this log, because there is a difference here.

I now understand why it works on your ambit.

**Review by landas — APPROVED** [review]

**Comment by landas** [comment]
Thank you very much. This information and code was really usefull.

**Comment by soutade** [comment]
You're welcome ! I just tested the merged code and it seems to works fine.
